File: 엄마호랑이반/10주차 문제/최혜령/단어변환.py
from collections import deque
def solution(begin, target, words):
    if target not in words:
        return 0
    
    produced = [False for _ in range(len(words))]
    print(produced)
    
    queue = deque()
    queue.append((begin, 0))
    
    while queue:
        cw, cnt = queue.popleft()
        if cw == target:
            return cnt
        for i in range(len(words)):
            if not produced[i]:
                diff = 0
                for j in range(len(cw)):
                    if cw[j] != words[i][j]:
                        diff += 1
                if diff == 1:
                    produced[i] = True
                    queue.append((words[i], cnt+1))
            
    
    answer = 0
    return answer

File: 엄마호랑이반/10주차 문제/이다영/단어 변환.java
package temp_dev;

import java.util.*;
import java.io.*;

public class Programmers {
	public static void main(String[] args) {
		System.out.println(solution("hit", "cog", new String[] {"hot", "dot", "dog", "lot", "log", "cog"} ));
		System.out.println(solution("hit", "cog", new String[] {"hot", "dot", "dog", "lot", "log"} ));
		
	}
	
	public static class Info {
		String word;
		int cnt;
		
		Info(String word, int cnt) {
			this.word = word;
			this.cnt = cnt;
		}
	}
	
	public static int solution(String begin, String target, String[] words) {
		if(!Arrays.asList(words).contains(target)) {
			return 0;
		}
		
		int res = Integer.MAX_VALUE;
		Deque<Info> dq = new ArrayDeque<>();
		
		for (int i = 0; i < words.length; i++) {
			if (isChange(begin, words[i])) {
				boolean[] used = new boolean[words.length];
				used[i] = true;
				dq.addLast(new Info(words[i], 1));
				
				while(!dq.isEmpty()) {
					Info info = dq.pollFirst();
					
					if (target.equals(info.word)) {
						res = Math.min(res, info.cnt);
					}
					
					for (int j = 0; j < words.length; j++) {
						if (!used[j] && isChange(info.word, words[j])) {
							used[j] = true;
							dq.addLast(new Info(words[j], info.cnt + 1));
						}
					}
				}
				
			}
		}
		
		return res > 51 ? 0 : res;
	}
	
	public static boolean isChange(String before, String after) {
		String[] bf = before.split("");
		for (String str : bf) {
			after = after.replaceFirst(str, "");
		}
		
		return after.length() == 1 ? true : false;
	}
}
